src/NBitcoin/Block.cs

```diff
@@ -241,23 +258,36 @@ public Block()
             this.SetNull();
         }
 
+        public Block(NetworkOptions networkOptions)
+            :this(new BlockHeader(networkOptions))
+        {            
+        }
+
         public Block(BlockHeader blockHeader)
         {
-            this.SetNull();
             this.header = blockHeader;
         }
 
-        public Block(byte[] bytes)
+        public Block(byte[] bytes, NetworkOptions networkOptions = null)
+            :this(networkOptions)
         {
-            this.ReadWrite(bytes);
+            this.ReadWrite(bytes, options:this.GetNetworkOptions());
         }
 
+        public NetworkOptions GetNetworkOptions()
+        {
+            return this.header.GetNetworkOptions();
+        }
 
         public void ReadWrite(BitcoinStream stream)
         {
+            if (stream.Serializing)
```

Comment:
I don't fully understand the reason behind this.

Reply:
When writing a block to the stream we would probably want any child objects that the block contains, and that will also be serialized, to use the same network options. Those child object may not have network options of their own and would be looking at the stream to provide the network options that dictate their serialization. That is the principle I am trying to follow consistently throughout.

Reply:
Hmm this is problematic as it can mean that there may be different options, I would qualify Options to be a fix think that does not change.

Ahh but it might happen that options are changed, like witness stuff that can flip the flags.
Hmm are we mixing behaviours types here?

Reply:
At a given hierarchical level the options do not change. If we have to temporarily change the options to perform some operation it's temporary for that one operation or any level below that. No long term effect higher up due to cloning.

Reply:
At a given hierarchical level the options do not change. If we have to temporarily change the options to perform some operation it's temporary for that one operation or any level below that. No long term effect higher up due to cloning.